src/pds/registrysweepers/reindexer/main.py

```diff
@@ -0,0 +1,371 @@
+
+
+def get_docs_query(filter_to_harvested_before: datetime):
+    """
+    Return a query to get all docs which haven't been reindexed by this sweeper and which haven't been harvested
+    since this sweeper process instance started running
+    """
+    # TODO: Remove this once query_registry_db_with_search_after is modified to remove mutation side-effects
+    return {
+        "query": {
+            "bool": {
+                "must_not": [{"exists": {"field": REINDEXER_FLAG_METADATA_KEY}}],
+                "must": {
+                    "range": {
+                        "ops:Harvest_Info/ops:harvest_date_time": {
+                            "lt": filter_to_harvested_before.astimezone(timezone.utc).isoformat()
+                        }
+                    }
+                },
+            }
+        }
+    }
+
+
+def fetch_dd_field_types(client: OpenSearch) -> Dict[str, str]:
+    dd_index_name = resolve_multitenant_index_name("registry-dd")
+    name_key = "es_field_name"
+    type_key = "es_data_type"
+    dd_docs = query_registry_db_with_search_after(
+        client,
+        dd_index_name,
+        _source={"includes": ["es_field_name", "es_data_type"]},
+        query={"query": {"match_all": {}}},
+        sort_fields=[name_key],
+    )
+    doc_sources = iter(doc["_source"] for doc in dd_docs)
+    dd_types = {
+        source[name_key]: source[type_key] for source in doc_sources if name_key in source and type_key in source
+    }
+    return dd_types
+
+
+def get_mapping_field_types_by_field_name(client: OpenSearch, index_name: str) -> Dict[str, str]:
+    return {
+        k: v["type"] for k, v in client.indices.get_mapping(index_name)[index_name]["mappings"]["properties"].items()
+    }
+
+
+def accumulate_missing_mappings(
+    dd_field_types_by_name: Dict[str, str], mapping_field_types_by_field_name: Dict[str, str], docs: Iterable[dict]
+) -> Dict[str, str]:
+    """
+    Iterate over all properties of all docs, test whether they are present in the given set of mapping keys, and
+    return a mapping of the missing properties onto their types.
+    @param dd_field_types_by_name: a mapping of document property names onto their types, derived from the data-dictionary db data
+    @param mapping_field_types_by_field_name: a mapping of document property names onto their types, derived from the existing index mappings
+    @param docs: an iterable collection of product documents
+    """
+
+    # TODO: map special-case property names onto their types, and incorporate them into the type resolution
+    #  NoneType indicates that the property is to be excluded
+    #  Anything with prefix 'ops:Provenance' should be excluded, as these properties are the responsibility of their
+    #  respective sweepers.
+    special_case_property_types_by_name = {
+        '@timestamp': None,
+        '@version': None,
+        '_package_id': None,
+        'description': 'text',
+        # 'lid':,
+        # 'lidvid',
```

Comment:
identifiers should be searchable from like queries (e.g. `*lidsubstring*`)

Reply:
https://github.com/NASA-PDS/registry-sweepers/pull/149#issuecomment-2436378805